pkg/backend/spinnaker/backend.go

```diff
@@ -856,3 +868,204 @@ func (s *SpinClient) SavePipeline(pipelineJSON string) (*http.Response, error) {
+func (s *SpinClient) getNestedPipelinesNames(stages interface{}, pipeline map[string]interface{}) ([]string, error) {
+	pipelineNames := []string{}
+	for _, stage := range stages.([]interface{}) {
+		stage := stage.(map[string]interface{})
+		stagePipelineField, exists := stage["pipeline"]
+		if !exists {
+			continue
+		}
+
+		stageTypeField, exists := stage["type"].(string)
+		if !exists || stageTypeField != "pipeline" {
+			continue
+		}
+
+		switch stagePipelineField.(type) {
+		case string:
+			continue
+		}
+
+		childPipeline := stage["pipeline"].(map[string]interface{})
+
+		if err := s.isValidPipeline(childPipeline); err != nil {
+			return pipelineNames, err
+		}
+
+		parentPipelineApplication := []string{pipeline["application"].(string)}
+
+		if err := isValidPipelineApplication(childPipeline, parentPipelineApplication); err != nil {
+			return pipelineNames, err
+		}
+
+		childPipelineStages, exists := childPipeline["stages"]
+		if exists {
+			hasChildPipelines, err := hasValidChildPipelineStages(childPipelineStages.([]interface{}), parentPipelineApplication)
+			if err != nil {
+				return pipelineNames, err
+			}
+
+			// If any of stages is of type pipeline get those pipelines names recursively
+			if hasChildPipelines {
+				nestedPipelineNames, err := s.getNestedPipelinesNames(childPipelineStages, childPipeline)
+				if err != nil {
+					return pipelineNames, err
+				}
+				pipelineNames = append(pipelineNames, nestedPipelineNames...)
+			}
+		}
+
+		// After we return from recursion we add "this layer" child pipeline
+		childPipelineName := childPipeline["name"].(string)
+		pipelineNames = append(pipelineNames, childPipelineName)
+	}
+
+	return pipelineNames, nil
+}
+
+func (s *SpinClient) getPipelinesNames(pipeline map[string]interface{}) ([]string, error) {
+	pipelineNames := []string{}
+
+	if stages, exists := pipeline["stages"]; exists {
+
+		hasChildPipelines, err := hasValidChildPipelineStages(stages.([]interface{}), []string{pipeline["application"].(string)})
+		if err != nil {
+			return pipelineNames, err
+		}
+
+		// If any of stages is of type pipeline collect those pipelines names recursively
+		if hasChildPipelines {
+			pipelineNames, err = s.getNestedPipelinesNames(stages, pipeline)
+			if err != nil {
+				return pipelineNames, err
+			}
+		}
+	}
+
+	pipelineName := pipeline["name"].(string)
+	pipelineNames = append(pipelineNames, pipelineName)
+
+	deduplicatedPipelineNames := func(pipelineNames []string) []string {
+		keys := make(map[string]bool)
+		list := []string{}
+
+		for _, entry := range pipelineNames {
+			if _, value := keys[entry]; !value {
+				keys[entry] = true
+				list = append(list, entry)
+			}
+		}
+		return list
+	}(pipelineNames)
+
+	return deduplicatedPipelineNames, nil
+}
+
+// GetPipelinesNamesAndApplication - gets list of names of all pipelines and application name configured
+func (s *SpinClient) GetPipelinesNamesAndApplication(pipelineJSON string) ([]string, string, error) {
```

Comment:
Why not add a `recursive` flag to `GetPipeline`?

It seems like `getPipelineNames` is really a recursive fetch.
Additionally this work has been done in `diff` which we can leverage.

Reply:
In `diff` `fillPipelineMap` does something simillar indeed, but gets children pipelines' ids by calling `GetPipeline` (**executor API call**) to match pipelines' ids to their code, for further diffing of "desiredPipeline" (locally rendered pipeline with Ids added from executor) with current pipeline persisted in executor.

`GetPipelinesNamesAndApplication`does  not require executor call - it only depends on local render results, as such enabling "instantaneous" `dry-run` option. As such `DeletePipeline` doesn't try to delete pipelines by spinnaker IDs but using pipelines' names only that are sufficient to do the job. 
This also answers the question "Why not add a recursive flag to `GetPipeline`" in my opinion. It operates against executor API which I wanted to avoid in my implementation.  
___
* As such I don't see these 2 code bases too similar in their functionality so that it would be "cheap" to re-use.
* I agree on the idea that these 'diff-related' functions should be moved to spinnaker/backend.
* Thinking regarding `GetPipeline` recursive option - this would help to move `diff` implementation into backend. We then will have to expose `d.Backend.Diff` function. Are we good to add this?
* Making `GetPipeline` as it is right now to serve the same purpose as `GetPipelinesNamesAndApplication`  would drastically change its purpose (right now it's an RPC call in a nutshell). This though be more aligned with moving diff into backend effort as it would require creating some internal function that would do the recursive fetch. Such recursive fetch can be optionally "local only" as `GetPipelinesNamesAndApplication` is. 

What I don't like in this solution is that a single function would server 2 very different purposes - I have a bad hunch on this.